File: backend/src/main/java/com/tmk/vtcmanager/application/domain/maintenance/Maintenance.java
package com.tmk.vtcmanager.application.domain.maintenance;

import com.tmk.vtcmanager.application.domain.operation.CategorieOperation;
import com.tmk.vtcmanager.application.domain.operation.DetailMaintenance;
import com.tmk.vtcmanager.application.domain.partenaire.Partenaire;
import com.tmk.vtcmanager.application.domain.vehicule.Vehicule;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Maintenance {

    private Long id;
    private String type;
    private LocalDate datePrevue;
    private LocalDate dateEffectuee;
    private Integer dureeHeures;
    private String description;
    private Integer kilometrageAuMoment;
    private Integer kilometrageProchaine;
    private BigDecimal cout;
    /**
     * Qui a réalisé l'intervention. C'était un texte libre : deux orthographes
     * du même garage faisaient deux prestataires, et rien ne se regroupait. Le
     * partenaire est désormais choisi dans le répertoire.
     */
    private Partenaire partenaire;
    private MaintenanceStatus statut;
    private Vehicule vehicule;
    private CategorieOperation categorieType;
    private DetailMaintenance detailMaintenance;

    /**
     * Pourquoi l'intervention a été annulée, qui l'a décidé et quand. Une
     * annulation sans motif ne se distingue pas d'un oubli : six mois plus
     * tard, personne ne sait plus si la vidange a été jugée inutile ou
     * simplement laissée de côté.
     */
    private String motifAnnulation;
    private String annulePar;
    private LocalDateTime annuleLe;

    /**
     * Renseigné à la lecture seulement : faux si un arrêté — période close,
     * caisse comptée — interdit désormais de restaurer cette maintenance
     * annulée. Le client s'en sert pour ne pas proposer une action vouée au
     * refus.
     */
    private Boolean restaurable;

    public void initializeDefaults() {
        if (this.statut == null) this.statut = MaintenanceStatus.PLANIFIEE;
    }

    public void terminer(BigDecimal coutEffectif, LocalDate dateEffectuee) {
        this.statut = MaintenanceStatus.TERMINEE;
        this.dateEffectuee = dateEffectuee != null ? dateEffectuee : LocalDate.now();
        if (coutEffectif != null) {
            this.cout = coutEffectif;
        }
    }

    public void annuler(String motif, String auteur) {
        this.statut = MaintenanceStatus.ANNULEE;
        this.motifAnnulation = motif;
        this.annulePar = auteur;
        this.annuleLe = LocalDateTime.now();
    }

    /**
     * Rend une maintenance annulée à l'état planifié : l'intervention est de
     * nouveau à faire, en entier. Elle n'a jamais été exécutée — ni date
     * d'exécution ni coût à restituer. Le marquage d'annulation s'efface : le
     * motif ne vaudrait plus rien en face d'une intervention de nouveau au
     * programme.
     */
    public void restaurer() {
        this.statut = MaintenanceStatus.PLANIFIEE;
        this.motifAnnulation = null;
        this.annulePar = null;
        this.annuleLe = null;
    }

    /**
     * Rouvre une maintenance terminée : elle repart PLANIFIEE, sans date
     * d'exécution ni coût.
     *
     * <p>La complétion annulée n'a pas eu lieu — l'intervention est à refaire
     * en entier, elle doit donc réapparaître parmi les maintenances à planifier
     * et non parmi celles qu'on croirait en cours au garage.
     */
    public void reouvrir() {
        this.statut = MaintenanceStatus.PLANIFIEE;
        this.dateEffectuee = null;
        this.cout = null;
    }

    public boolean isPlanifiee() {
        return this.statut == MaintenanceStatus.PLANIFIEE;
    }
}

File: mobile/lib/features/maintenance/domain/entities/maintenance.dart
import '../../../operation_financiere/domain/entities/detail_maintenance.dart';

class Maintenance {
  final int? id;
  final String type;
  final DateTime datePrevue;
  final DateTime? dateEffectuee;
  final int? dureeHeures;
  final String? description;
  final int? kilometrageAuMoment;
  final int? kilometrageProchaine;
  final double? cout;
  /// Partenaire ayant réalisé l'intervention (référentiel des partenaires).
  final int? partenaireId;
  final String? partenaireNom;
  final String? statut;
  final int? vehiculeId;
  final String? vehiculeNom;
  final String? vehiculeImmatriculation;
  final int? categorieTypeId;
  final String? categorieTypeLibelle;
  final DetailMaintenance? detailMaintenance;

  /// Pourquoi l'intervention a été annulée, par qui et quand. Nul tant qu'elle
  /// est au programme.
  final String? motifAnnulation;
  final String? annulePar;
  final DateTime? annuleLe;

  /// Faux si un arrêté — période comptable close, caisse comptée — interdit
  /// désormais la restauration. Le bouton « Restaurer » est alors masqué :
  /// le serveur refuserait.
  final bool restaurable;

  const Maintenance({
    this.id,
    required this.type,
    required this.datePrevue,
    this.dateEffectuee,
    this.dureeHeures,
    this.description,
    this.kilometrageAuMoment,
    this.kilometrageProchaine,
    this.cout,
    this.partenaireId,
    this.partenaireNom,
    this.statut,
    this.vehiculeId,
    this.vehiculeNom,
    this.vehiculeImmatriculation,
    this.categorieTypeId,
    this.categorieTypeLibelle,
    this.detailMaintenance,
    this.motifAnnulation,
    this.annulePar,
    this.annuleLe,
    this.restaurable = false,
  });

  bool get isPending => statut == 'PLANIFIEE' || statut == null;
  bool get isDone => statut == 'TERMINEE';

  /// Une intervention close ne se retouche plus : terminée, son coût et sa date
  /// sont ceux de la dépense déjà passée au journal ; annulée, elle n'a plus
  /// rien à décrire. Le formulaire d'édition leur est fermé — c'est une
  /// nouvelle intervention qu'il faut planifier.
  bool get estModifiable => statut != 'TERMINEE' && statut != 'ANNULEE';

  Maintenance copyWith({
    int? id,
    String? type,
    DateTime? datePrevue,
    DateTime? dateEffectuee,
    int? dureeHeures,
    String? description,
    int? kilometrageAuMoment,
    int? kilometrageProchaine,
    double? cout,
    int? partenaireId,
    String? partenaireNom,
    String? statut,
    int? vehiculeId,
    String? vehiculeNom,
    String? vehiculeImmatriculation,
    int? categorieTypeId,
    String? categorieTypeLibelle,
    DetailMaintenance? detailMaintenance,
  }) {
    return Maintenance(
      id: id ?? this.id,
      type: type ?? this.type,
      datePrevue: datePrevue ?? this.datePrevue,
      dateEffectuee: dateEffectuee ?? this.dateEffectuee,
      dureeHeures: dureeHeures ?? this.dureeHeures,
      description: description ?? this.description,
      kilometrageAuMoment: kilometrageAuMoment ?? this.kilometrageAuMoment,
      kilometrageProchaine: kilometrageProchaine ?? this.kilometrageProchaine,
      cout: cout ?? this.cout,
      partenaireId: partenaireId ?? this.partenaireId,
      partenaireNom: partenaireNom ?? this.partenaireNom,
      statut: statut ?? this.statut,
      vehiculeId: vehiculeId ?? this.vehiculeId,
      vehiculeNom: vehiculeNom ?? this.vehiculeNom,
      vehiculeImmatriculation:
          vehiculeImmatriculation ?? this.vehiculeImmatriculation,
      categorieTypeId: categorieTypeId ?? this.categorieTypeId,
      categorieTypeLibelle: categorieTypeLibelle ?? this.categorieTypeLibelle,
      detailMaintenance: detailMaintenance ?? this.detailMaintenance,
      // Le marquage d'annulation et la restaurabilité ne se saisissent pas :
      // ils viennent du serveur et suivent la copie, sinon une maintenance
      // annulée perdrait son motif au premier copyWith.
      motifAnnulation: motifAnnulation,
      annulePar: annulePar,
      annuleLe: annuleLe,
      restaurable: restaurable,
    );
  }
}
